mender-update/deployments/platform/boost_log/deployments.cpp

```diff
@@ -0,0 +1,197 @@
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <mender-update/deployments.hpp>
+
+#include <boost/date_time/posix_time/posix_time.hpp>
+#include <boost/log/attributes.hpp>
+#include <boost/log/common.hpp>
+#include <boost/log/sinks.hpp>
+#include <boost/smart_ptr/shared_ptr.hpp>
+
+#include <filesystem>
+#include <string>
+
+#include <common/error.hpp>
+#include <common/io.hpp>
+#include <common/json.hpp>
+#include <common/log.hpp>
+#include <common/path.hpp>
+
+namespace mender {
+namespace update {
+namespace deployments {
+
+using namespace std;
+
+namespace logging = boost::log;
+namespace expr = boost::log::expressions;
+namespace sinks = boost::log::sinks;
+
+namespace fs = std::filesystem;
+
+namespace error = mender::common::error;
+namespace io = mender::common::io;
+namespace json = mender::common::json;
+namespace mlog = mender::common::log;
+namespace path = mender::common::path;
+
+static void JsonLogFormatter(logging::record_view const &rec, logging::formatting_ostream &strm) {
+	strm << "{";
+
+	auto val = logging::extract<boost::posix_time::ptime>("TimeStamp", rec);
+	if (val) {
+		strm << R"("timestamp":")"
+			 << json::EscapeString(boost::posix_time::to_iso_extended_string(val.get())) << "\",";
+	}
+
+	auto level = logging::extract<mlog::LogLevel>("Severity", rec);
+	if (level) {
+		string lvl = mlog::ToStringLogLevel(level.get());
+		strm << R"("level":")" << json::EscapeString(lvl) << "\",";
+	}
+
+	strm << R"("message":")" << json::EscapeString(*rec[expr::smessage]) << "\"}";
+}
+
+static const size_t kMaxExistingLogs = 5;
+static const uintmax_t kLogsFreeSpaceRequired = 100 * 1024; // 100 KiB
+
+error::Error DeploymentLog::PrepareLogDirectory() {
+	fs::path dir_path(data_store_dir_);
+	bool created = fs::create_directories(dir_path);
+	if (created) {
+		// should basically never happen, but if we happened to create the
+		// directory, it's empty and thus well-prepared
+		return error::NoError;
+	}
+
+	vector<string> old_logs;
+	fs::directory_iterator end_itr; // default construction yields past-the-end
+	for (fs::directory_iterator itr(dir_path); itr != end_itr; ++itr) {
+		fs::path file_path = itr->path();
+		if (!fs::is_regular_file(file_path)) {
+			continue;
+		}
+
+		string file_name = file_path.filename().string();
+
+		if (file_name == "deployments.0000." + id_ + ".log") {
+			// this log file will be (re)used, leave it alone
+			continue;
+		}
+
+		if ((file_name.find("deployments.") != 0)
+			|| (file_name.substr(file_name.size() - 4) != ".log")) {
+			continue;
+		}
+
+		old_logs.push_back(file_name);
+	}
+	std::sort(old_logs.begin(), old_logs.end());
+
+	error_code ec;
+	fs::space_info space_info = fs::space(dir_path, ec);
+	if (ec) {
+		return error::Error(
+			ec.default_error_condition(), "Failed to check free space for log files");
+	}
+
+	while ((old_logs.size() > 0)
```

Comment:
This could end up deleting *all* files, can't it? But we don't want that in the "same ID / append" case.

Reply:
That file is excluded from `old_logs`.